### DISCUSSION TITLE
Why does it require exactly 4MB of RAM?

### DISCUSSION THREAD

**Discussion by IamShodan** [vvaltchev/tilck#78] [Q&A] [answered]
This looks like AMAZING project for my 386SX 25Mhz with 2MB of RAM.

Tell me - why does it require exactly 4MB of ram, even if kernel size itself its much much smaller?
For example in old Linux kernels (before 2.0 for sure) we could run Linux at just 2 MB of ram - but then SOMETHING happened and such feature is gone - 4MB of ram if hard-coded in the kernel.

Do you know why they changed it and how to bring it back? Or even to remove any hard limit whatsoever and just TRY to boot up

ps. I suspect this something about 4MB memory pages.

**Comment by vvaltchev** [comment, marked as answer]
Hi!
Thanks for taking interest in the project!

Tilck does **not** require at least 4 MB of RAM. Theoretically, it could run with 2 MB as well. Just, I couldn't properly test it with 2 MB because `qemu-system-i386` doesn't support less than 4 MB of memory. At least, it didn't in the past.

Now I tried again, and I just discovered that QEMU 4.2 supports running a x86 machine with less than 4 MB of RAM. So, I **confirm** that Tilck runs just fine with **3 MB** of RAM (if properly configured for the smallest scale, of course). 

With 2 MB of RAM, it **doesn't**. I believe the kernel could run, but it would require a very heavily stripped down version of busybox and an overall much smaller `initrd` (currently, 540 KB are way too much). Plus, maybe some additional tweaking in the kernel itself. So, with the current setup, 2 MB is not enough. (Consider that not even the whole 2 MB RAM cannot be used, as there are parts of it reserved.)

Now, about **386SX**: unfortunately, Tilck won't run there not because of the RAM, but because it requires the **i586** instruction set. That's because it uses 4-MB pages and several other features and instructions not available on CPUs older than i586. Since I don't have such older machine to test Tilck, I didn't even try supporting a special configuration for them. Maybe in the long term, I will just for the fun of it. But that would require also an USB floppy drive + floppies on which I'll write Tilck's image. Floppies are not produced anymore, as far as I know. In other words, it would be challenging today to have such a setup.

About **Linux**, I believe it can run on i486 if properly configured, but not on anything older than that. Support for i386 has been dropped in 2012: https://www.zdnet.com/article/good-bye-386-linux-to-drop-support-for-i386-chips-with-next-major-release/

Vlad

**Reply by IamShodan** [reply]
Thank you for quick response!

As for older hardware without network cards - you don't need floppies to boot. You could simply use SLIP/PPP over ther serial cable(with MS-DOS) to transfer image to the hard-drive, and then use something like loadlin.exe or custom boot-loader directly from MS-DOS environment. 

[https://medium.com/swlh/connecting-a-286-dos-pc-to-the-internet-through-a-serial-connection-in-2019-b93a422ff094](https://medium.com/swlh/connecting-a-286-dos-pc-to-the-internet-through-a-serial-connection-in-2019-b93a422ff094)

I remember that I was able to run Slackware 2 or 3.1 with just 2mb of RAM, same goes to  Minix, that's why I keep asking this question.

ps. If you want REAL emulator (much better than qemu!) for older harware I strongly suggest trying ["PCEM". ](https://pcem-emulator.co.uk/). 

It's like QEMU but it emulate actual speed of hardware too. Super useful!

**Reply by vvaltchev** [reply]
Good point! Just, I'd need an MS-DOS diskette in that case, or to have it installed on disk somehow. But still, it's doable. There are external IDE adapters etc. But, I guess, I'd have to first buy a house before starting to create a nice collection of legacy machines.

Wow man, you used Slackware 2! Happy to meet such a Linux veteran :-) 

PCEM looks great, thanks for the hint. I didn't know about it. I would definitely use it to test Tilck on older hardware.
